### PR TITLE
[Security] Make security.providers optional

### PR DESCRIPTION
**Pull request by MatTheCat**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #21998
| License       | MIT

Don't really know if it's viable but I just hit #21998 so I would like to tackle this.

### CODE REVIEW

**Comment by MatTheCat** [comment]
@chalasr @stof  do you think this can cause issues?

Also I don't know how to provide a proper error message in case a provider is needed but none is configured.

**Comment by chalasr** [comment]
Very complex topic, looks like you like them :) (#24805)

Me and @weaverryan have this in mind for a while, the issue is that actually pretty much all security listeners are requiring a user provider for user loading/refreshment.
This misses new test case(s) (authentication with no configured provider), adding them will better answer your question.

**Comment by MatTheCat** [comment]
Yes the way we'll authenticate our user makes me discover some edge cases!

> the issue is that actually pretty much all security listeners are requiring a user provider for user loading/refreshment.

I don't understand how this is an issue?

**Comment by MatTheCat** [comment]
> This misses new test case(s) (authentication with no configured provider), adding them will better answer your question.

Isn't this covered by `testMissingProviderForListener`?

File: src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/SecurityExtensionTest.php [hunk 1]
```diff
@@ -166,8 +166,7 @@ public function testPerListenerProvider()
     }
 
     /**
-     * @expectedException \Symfony\Component\Config\Definition\Exception\InvalidConfigurationException
-     * @expectedExceptionMessage Not configuring explicitly the provider for the "http_basic" listener on "ambiguous" firewall is ambiguous as there is more than one registered provider.
+     * @expectedException \TypeError
```
Comment (chalasr) [inline src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/SecurityExtensionTest.php]: getting a `TypeError` when not setting the provider isn't fine to me as this aims to make it optional

Comment (MatTheCat) [reply src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/SecurityExtensionTest.php]: Yes I agree we need a proper error but I don't know when to throw it.

**Comment by MatTheCat** [comment]
What if instead of `null` we set the provider as a dummy user provider which would throw an exception in its constructor?

**Comment by MatTheCat** [comment]
I created a dummy user provider whose sole purpose is to throw an exception, WDYT?

**Comment by weaverryan** [comment]
Huh, a dummy provider is a really clever idea - it makes this very feasible (and sure, we could try to rip out the guts of the user provider system later... but that will be super hard and will probably never happen).

I like this!

**Comment by MatTheCat** [comment]
Do we need a functional test where authentication is done without user provider?

**Comment by MatTheCat** [comment]
Are we good to go then or are some things missing?

File: src/Symfony/Bundle/SecurityBundle/Tests/Functional/Bundle/MissingUserProviderBundle/MissingUserProviderBundle.php [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\SecurityBundle\Tests\Functional\Bundle\MissingUserProviderBundle;
+
+use Symfony\Component\HttpKernel\Bundle\Bundle;
+
+class MissingUserProviderBundle extends Bundle
```
Comment (chalasr) [inline src/Symfony/Bundle/SecurityBundle/Tests/Functional/Bundle/MissingUserProviderBundle/MissingUserProviderBundle.php]: I would mark it `@internal`

**Comment by MatTheCat** [comment]
Is it alright to let `BadMethodCallException`s empty?

File: src/Symfony/Component/Security/Core/User/MissingUserProvider.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+{
+    /**
+     * @param string $firewall the firewall missing a provider
+     */
+    public function __construct($firewall)
+    {
+        throw new InvalidConfigurationException(sprintf(
+            '"%s" firewall requires a user provider but none was defined.',
+            $firewall
+        ));
```
Comment (fabpot) [inline src/Symfony/Component/Security/Core/User/MissingUserProvider.php]: small thing: should be on one line.

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php [hunk 4]
```diff
@@ -422,6 +422,12 @@ private function createAuthenticationListeners($container, $id, $firewall, &$aut
                         $userProvider = null;
                     } elseif ($defaultProvider) {
                         $userProvider = $defaultProvider;
+                    } elseif (empty($providerIds)) {
+                        $userProvider = sprintf('security.user.provider.missing.%s', $key);
+                        $container->setDefinition(
+                            $userProvider,
+                            (new ChildDefinition('security.user.provider.missing'))->replaceArgument(0, $id)
+                        );
```
Comment (fabpot) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: CS fix: should be on one line.

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php (hunk 4 above)
Comment (MatTheCat) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: Oh okay.

File: src/Symfony/Component/Security/Core/User/MissingUserProvider.php [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+ * when a firewall requires a user provider but none was defined.
+ *
+ * @internal
+ */
+class MissingUserProvider implements UserProviderInterface
+{
+    /**
+     * @param string $firewall the firewall missing a provider
+     */
+    public function __construct($firewall)
```
Comment (chalasr) [inline src/Symfony/Component/Security/Core/User/MissingUserProvider.php]: string typehint

**Review by chalasr — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @MatTheCat.

**Pull request merged**